### PR TITLE
Add support for UTF-8 decoding

### PR DESCRIPTION
**Pull request by ftruzzi**
This PR adds support for UTF-8 password decoding as discussed in https://github.com/jaraco/keyring/issues/438

Wondering if we should add a warning when UTF-16 decoding fails? Something like "you can read this password but won't be able to write it back in the same format using this library at this time"?

Thanks!

### CODE REVIEW

**Comment by jaraco** [comment]
I've taken the liberty to refactor the behavior into a `DecodingCredential` class, consolidating the behavior.

> Wondering if we should add a warning when UTF-16 decoding fails?

That seems reasonable, especially because a warning could be suppressed by an environment variable.

One additional thing, could you:

- [ ] Create a test to capture the new expectation? The test will probably need to inject a UTF-8 password into the credential store (and make sure it's registered as created in order to be deleted).

**Comment by ftruzzi** [comment]
Thanks! I have added a warning. I wanted to add a test from the beginning, but how can we write a test UTF-8 credential to the credential store if `pywin32-ctypes` doesn't support it?

**Comment by jaraco** [comment]
> Thanks! I have added a warning. I wanted to add a test from the beginning, but how can we write a test UTF-8 credential to the credential store if `pywin32-ctypes` doesn't support it?

Good question. I guess it's not easy. You'd need to find a back door to add the credential, either another executable or perhaps invoking the underlying API directly (through ctypes).

Now that I re-read through the issue, I see we agreed to support reading these before writing them, so perhaps having a test would require at least nominal support for writing them. And clearly, using `Slack` or another electron app is probably not a very portable solution.

So how about let's for now just write a stub for a test with a skip indicating that the functionality isn't available? Something like:

```
@pytest.mark.skip(reason="writing utf-8 isn't yet supported #438")
def test_utf_8_credential(self):
    # todo: save a non-ascii password in UTF-8
    pw = self.keyring.get_password(system, username)
    assert isinstance(pw, str)
```

**Comment by ftruzzi** [comment]
> You'd need to find a back door to add the credential, either another executable or perhaps invoking the underlying API directly (through ctypes).

Thanks! It looks like manually editing the ctypes credential worked :) It can also be a first step towards supporting it officially if you like the hacky solution. Can you take a look at the new test?

Sorry, realizing this just now:
> perhaps having a test would require at least nominal support for writing them.

Up to you then, we can use a stub test for this and consider better support later, as you prefer

**Comment by jaraco** [comment]
> > You'd need to find a back door to add the credential, either another executable or perhaps invoking the underlying API directly (through ctypes).
> 
> Thanks! It looks like manually editing the ctypes credential worked :) It can also be a first step towards supporting it officially if you like the hacky solution. Can you take a look at the new test?

Great!

> > perhaps having a test would require at least nominal support for writing them.
> 
> Up to you then, we can use a stub test for this and consider better support later, as you prefer

No problem. Looks like were close here, and having a working test is best.

I tried applying this patch to ensure that unicode characters get represented correctly:

```diff
diff --git a/tests/backends/test_Windows.py b/tests/backends/test_Windows.py
index 0c63595..645ed89 100644
--- a/tests/backends/test_Windows.py
+++ b/tests/backends/test_Windows.py
@@ -3,7 +3,7 @@ import sys
 import pytest
 
 import keyring.backends.Windows
-from keyring.testing.backend import BackendBasicTests
+from keyring.testing.backend import BackendBasicTests, UNICODE_CHARS
 
 
 @pytest.mark.skipif(
@@ -56,7 +56,7 @@ class TestWinVaultKeyring(BackendBasicTests):
         """
         service = "keyring-utf8-test"
         username = "keyring"
-        password = "utf8-test"
+        password = UNICODE_CHARS
 
         self.set_utf8_password(service, username, password)
         assert self.keyring.get_password(service, username) == password
```

But when I did, the test failed in the final assertion:

```
_____________________________________________ TestWinVaultKeyring.test_read_utf8_password ______________________________________________ 

self = <test_Windows.TestWinVaultKeyring object at 0x000001D80CB63910>

    def test_read_utf8_password(self):
        """
        Write a UTF-8 encoded credential and make sure it can be read back correctly.
        """
        service = "keyring-utf8-test"
        username = "keyring"
        password = UNICODE_CHARS
     
        self.set_utf8_password(service, username, password)
>       assert self.keyring.get_password(service, username) == password
E       AssertionError: assert '雗铗鯗駗ꏗꇗ\uaad7...胑諑럐뷐냐뫐냐苑뻐돐賑뻐뷐' == 'זהכיףסתםלשמו...ръзнакатогьон'
E         - זהכיףסתםלשמועאיךתנצחקרפדעץטובבגןξεσκεπάζωτηνψυχοφθόραβδελυγμίαСъешьжеещёэтихмягкихфранцузскихбулокдавыпейчаюЖълтатадюлябешещастливачепухъткойтоцъфназамръзнакатогьон
E         + 雗铗鯗駗ꏗꇗ꫗鷗鳗꧗黗闗ꋗ郗駗髗꫗ꃗꛗ韗ꟗ꣗ꓗ鏗ꋗꗗ飗闗釗釗鋗鿗뻎뗎菏뫎뗎胏곎뛎觏蓏럎뷎裏藏蟏뿎蛏룎賏臏뇎닎듎뗎믎藏돎볎꿎뇎ꇐ諑뗐裑賑뛐뗐뗐觑金跑苑룐藑볐近돐뫐룐藑蓑胑냐뷐蛑菑럐臑뫐룐藑뇐菑믐뻐뫐듐냐닐译뿐뗐말蟑냐軑雐諑믐苑냐苑냐듐軑믐近뇐뗐裑뗐觑냐臑苑믐룐닐냐蟑뗐뿐菑藑諑苑뫐뻐말苑뻐蛑諑蓑뷐냐럐냐볐胑諑럐뷐냐뫐냐苑뻐돐賑뻐뷐

tests\backends\test_Windows.py:62: AssertionError
```

It appears as if some UTF-8 can be decoded as UTF-16.

**Comment by jaraco** [comment]
Adding just some ascii characters to the password and the test passes, even with the rest of those unicode characters present, so I think that confirms my suspicion that UNICODE_CHARS.encode('utf-8') is also valid UTF-16.

Still, I guess that's an acceptable limitation.

**Comment by ftruzzi** [comment]
Good! I also came to the same conclusion.

Looks like there are no UTF-16 byte order marks in the credentials saved with ctypes so I don't think there's an easy way out of this. Maybe detecting UTF-8 with `chardet` and falling back to UTF-16 would work but that would be an extra dependency and I don't know enough about encodings to know how frequent these collisions are, let alone in passwords!
